Fixing types and optionality

7jrxt42BxFZo4iAnN4CX opened the pull request:
- Added a debug flag feature, which displays additional information about the field on which the deserialization error occurred.
- Fixing incorrect types
- Fixing optional fields for #68
- Fixes for passing the test_api_responses_missing_fields test
- Added test function (test_get_quote_period_interval) for #62
- The blocking version has been updated too

Commit:
Update Cargo.toml, async_impl.rs, blocking_impl.rs, and 3 more files

cjglo commented on src/blocking_impl.rs:
```diff
@@ -284,14 +277,14 @@ mod tests {
         assert_eq!(&response.chart.result[0].meta.symbol, "BF.B");
-        assert_eq!(&response.chart.result[0].meta.range, "1d");
-        assert_eq!(&response.chart.result[0].meta.data_granularity, "1m");
+        // assert_eq!(&response.chart.result[0].meta.range, "1d");
```
Might as well delete these lines, unless there is a reason to keep them there and commented out?
```suggestion
```

7jrxt42BxFZo4iAnN4CX replied:
I just forgot.
asset "BF.B" has no quote data

cjglo commented on src/blocking_impl.rs:
```diff
@@ -284,14 +277,14 @@ mod tests {
-        assert_eq!(&response.chart.result[0].meta.range, "1d");
-        assert_eq!(&response.chart.result[0].meta.data_granularity, "1m");
+        // assert_eq!(&response.chart.result[0].meta.range, "1d");
+        // assert_eq!(&response.chart.result[0].meta.data_granularity, "1m");
```
```suggestion
```

cjglo commented:
Also this does bring up a good point, I think you mentioned this in the issue, but going forward I assume almost every field should be optional going forward?  I believe the actual Yahoo API has every field as optional (but for obvious reasons, that doesn't come up in python).

7jrxt42BxFZo4iAnN4CX committed:
Update src/blocking_impl.rs

Delete unnecessary comment

Co-authored-by: Chris J G <62271962+cjglo@users.noreply.github.com>

7jrxt42BxFZo4iAnN4CX committed:
Update src/blocking_impl.rs

Delete unnecessary comment

Co-authored-by: Chris J G <62271962+cjglo@users.noreply.github.com>

7jrxt42BxFZo4iAnN4CX commented:
> Also this does bring up a good point, I think you mentioned this in the issue, but going forward I assume almost every field should be optional going forward? I believe the actual Yahoo API has every field as optional (but for obvious reasons, that doesn't come up in python).

I think there are exceptions, but in most cases everything is optional (different assets and different exchanges have different data). But I just fix the errors that I catch when using (To just have objects, not errors). You can probably change all u64 fields to i64, so that there are definitely no problems with negative numbers.

p.s. I didn't look at the Python library, I just found the necessary information through a GitHub search(code search).

Commit:
Improved error handling

7jrxt42BxFZo4iAnN4CX commented:
**Improved error handling:**

request: https://query1.finance.yahoo.com/v8/finance/chart/AAPL?symbol=AAPL&period1=1735718158&period2=1740988558&interval=5m&events=div%7Csplit%7CcapitalGains&includePrePost=true

response:
```json
{
"chart":{
   "result":null,
    "error": {
         "code":"Unprocessable Entity",
         "description":"5m data not available for startTime=1735718158 and endTime=1740988558. The requested range must be within the last 60 days."
        }
    }
}
```
such an answer is now transformed into:
```rust 
Err(ApiError(YErrorMessage { code: Some("Unprocessable Entity"), description: Some("5m data not available for startTime=1735747158 and endTime=1741017558. The requested range must be within the last 60 days.") }))
```

@xemwebe when will there be time for a merge?

7jrxt42BxFZo4iAnN4CX commented:
or another example:
```rust 
ApiError(YErrorMessage { code: Some("Not Found"), description: Some("No data found, symbol may be delisted") })
```

Commit:
Improved f64 deserialization: support for "NaN" value.
Fixes deserialization errors.

Commit:
fix blocking test

Commit:
fix blocking test

7jrxt42BxFZo4iAnN4CX commented:
fixes:

> DeserializeFailed(Error` { path: Path { segments: [Map { key: "quoteSummary" }, Map { key: "result" }, Seq { index: 0 }, Map { key: "defaultKeyStatistics" }, Map { key: "forwardPE" }] }, original: Error("invalid type: string \"Infinity\", expected f64", line: 1, column: 3798) })

> DeserializeFailed(Error { path: Path { segments: [Map { key: "quoteSummary" }, Map { key: "result" }, Seq { index: 0 }, Map { key: "summaryDetail" }, Map { key: "trailingAnnualDividendYield" }] }, original: Error("invalid type: string \"NaN\", expected f64", line: 1, column: 3801) })

> DeserializeFailed(Error { path: Path { segments: [Map { key: "quoteSummary" }, Map { key: "result" }, Seq { index: 0 }, Map { key: "summaryDetail" }, Map { key: "priceToSalesTrailing12Months" }] }, original: Error("invalid type: string \"Infinity\", expected f64", line: 1, column: 3567) })

> DeserializeFailed(Error { path: Path { segments: [Map { key: "quoteSummary" }, Map { key: "result" }, Seq { index: 0 }, Map { key: "defaultKeyStatistics" }, Map { key: "forwardPE" }] }, original: Error("invalid type: string \"Infinity\", expected f64", line: 1, column: 3798) })

> DeserializeFailed(Error { path: Path { segments: [Map { key: "quoteSummary" }, Map { key: "result" }, Seq { index: 0 }, Map { key: "quoteType" }, Map { key: "firstTradeDateEpochUtc" }] }, original: Error("invalid value: integer `-252322200`, expected u64", line: 1, column: 6531) })

> DeserializeFailed(Error { path: Path { segments: [Map { key: "quoteSummary" }, Map { key: "result" }, Seq { index: 0 }, Map { key: "defaultKeyStatistics" }, Map { key: "lastSplitDate" }] }, original: Error("invalid value: integer `-226281600`, expected u64", line: 1, column: 6076) })

> DeserializeFailed(Error { path: Path { segments: [Map { key: "quoteSummary" }, Map { key: "result" }] }, original: Error("invalid type: null, expected a sequence", line: 1, column: 30) })

Commit:
removed unnecessary check_consistency

Commit:
fix test

7jrxt42BxFZo4iAnN4CX commented:
* Renamed check_consistency to check_historical_consistency.
* Removed check_historical_consistency where it is not needed.

Now if the asset has no historical data, we can get metadata if it is. And we will get the error YahooError::NoQuotes if we try to call last_quote() or quotes().

Commit:
FIX: key: "regularMarketTime" , Error("invalid type: null, expected u32"}

Commit:
make map_error_msg pub(crate)

xemwebe commented:
Ist this branch still under development or is it ready for merger from your perspective (though one unit test fails, it seems).

7jrxt42BxFZo4iAnN4CX commented:
It is ready for merging. The error occurred in a part of the code that I did not touch (search for options), perhaps something changed in the API (did not look).

7jrxt42BxFZo4iAnN4CX commented:
https://query2.finance.yahoo.com/v6/finance/options/AAPL

>Yahoo
Will be right back...
Thank you for your patience.
Our engineers are working quickly to resolve the issue.

dancixx commented:
@7jrxt42BxFZo4iAnN4CX Can you please consider modifying the fields of the single quote struct to optional:

```rust
/// Struct for single quote
#[derive(Debug, Clone, PartialEq, PartialOrd, Deserialize, Serialize)]
pub struct Quote {
    pub timestamp: u64,
    pub open: Decimal,
    pub high: Decimal,
    pub low: Decimal,
    pub volume: u64,
    pub close: Decimal,
    pub adjclose: Decimal,
}
```

I am getting this error:
```
Error: deserializing response from yahoo! finance failed

Caused by:
    missing field `adjclose`
```

I thought of creating a PR, but I found yours, which is related to this problem.

7jrxt42BxFZo4iAnN4CX commented:
@dancixx It is quite possible that the problem has already been solved in this PR. Can you check on my fork or wait for the merge?

adjclose is also found in one more place:
```rust
#[derive(Deserialize, Debug)]
pub struct QuoteBlock {
    quote: Vec<QuoteList>,
    #[serde(default)]
    adjclose: Option<Vec<AdjClose>>,
}
```
